tree-walk-xl.go

```diff
@@ -138,23 +140,37 @@ func (xl xlObjects) treeWalk(bucket, prefixDir, entryPrefixMatch, marker string,
 				// recursing into "four/"
 				markerArg = markerBase
 			}
-			*count--
 			prefixMatch := "" // Valid only for first level treeWalk and empty for subdirectories.
-			if !xl.treeWalk(bucket, pathJoin(prefixDir, entry), prefixMatch, markerArg, recursive, send, count, isLeaf) {
-				return false
+			if i == len(entries)-1 && stackDepth == 0 {
+				isEnd = true
 			}
+			stackDepth++
+			xl.treeWalk(bucket, pathJoin(prefixDir, entry), prefixMatch, markerArg, recursive, isLeaf, treeWalkCh, doneCh, stackDepth, isEnd)
+			stackDepth--
 			continue
 		}
-		*count--
-		if !send(treeWalkResult{entry: pathJoin(prefixDir, entry)}) {
-			return false
+		var isEOF bool
+		if stackDepth == 0 && i == len(entries)-1 {
+			isEOF = true
+		} else if i == len(entries)-1 && isEnd {
+			isEOF = true
 		}
+		select {
+		case <-doneCh:
+			if stackDepth == 0 {
+				close(treeWalkCh)
+				return
+			}
+		case treeWalkCh <- treeWalker{entry: pathJoin(prefixDir, entry), end: isEOF}:
```

Comment:
> this go routine will block and there will be a goroutine leak if the timeout had happened when it was trying to send an entry in one of the subdirectories. that's why previously treeWalk was returning a value to detect if timeout had already happened or if there was an error.
> This won't block when a timeout occurs doneCh is closed i.e select receives a reply where the close(treeWalkCh) is issued and returned.

The select here is the replacement for 'send()' function which was present previously.